### PR TITLE
Snapshot store

### PR DESCRIPTION
**Pull request by joost-de-vries**
I've implemented support for snapshot storage. #56 

I had to refactor some shared logic from the `akka.persistence.dynamodb.journal` package to `akka.persistence.dynamodb`. I did not refactor the configuration properties for shared values. So currently properties for the snapshot are be default a copy of the journals values.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -54,6 +55,27 @@ Before you can use these settings you will have to create a table, e.g. using th
 
   * a hash key of type String with name `par`
   * a sort key of type Number with name `num`
+  
+### Snapshot
```
Comment (ktoso) [inline README.md]: `Snapshot Store` perhaps :)

File: src/main/resources/reference.conf [hunk 2]
```diff
@@ -121,3 +121,54 @@ dynamodb-journal {
...
+
+  aws-client-config =${dynamodb-journal.aws-client-config}
+
+  # AWS API limits - DO NOT CHANGE UNLESS YOU KNOW WHAT YOU ARE DOING
+  aws-api-limits {
+    max-batch-get = ${dynamodb-journal.aws-api-limits.max-batch-get}
+    max-batch-write = ${dynamodb-journal.aws-api-limits.max-batch-write}
+    max-item-size = ${dynamodb-journal.aws-api-limits.max-item-size}
+  }
+}
```
Comment (ktoso) [inline src/main/resources/reference.conf:171]: AFAIR to avoid problems with concating resources it's best to include an empty line after the last `}`

File: src/main/resources/reference.conf [hunk 3]
```diff
@@ -121,3 +121,54 @@ dynamodb-journal {
     }
   }
 }
+dynamodb-snapshot{
```
Comment (ktoso) [inline src/main/resources/reference.conf]: `dynamodb-snapshot-store`

File: src/main/resources/reference.conf [hunk 4]
```diff
@@ -121,3 +121,54 @@ dynamodb-journal {
...
+
+  class="akka.persistence.dynamodb.snapshot.DynamoDBSnapshotStore"
+
+  # The service endpoint to connect to for the DynamoDB instance that
+  # shall be used. Please refer to the AWS documentation for details.
+  endpoint = ${dynamodb-journal.endpoint}
```
Comment (ktoso) [inline src/main/resources/reference.conf:130]: good

File: README.md [hunk 5]
```diff
@@ -54,6 +55,27 @@ Before you can use these settings you will have to create a table, e.g. using th
...
+~~~
+akka.persistence.snapshot-store.plugin = "my-dynamodb-snapshot"
+
+my-dynamodb-snapshot = ${dynamodb-snapshot} # include the default settings
+my-dynamodb-snapshot {                     # and add some overrides
+    snapshot-table =  <the name of the table to be used>
+    journal-name =  <prefix to be used for all keys stored by this plugin>
+    aws-access-key-id =  <your key>
+    aws-secret-access-key =  <your secret>
+    endpoint =  "https://dynamodb.us-east-1.amazonaws.com" # or where your deployment is
```
Comment (ktoso) [inline README.md]: perhaps add a note here that defaults to same as journal

File: src/main/scala/akka/persistence/dynamodb/journal/DynamoDBJournalConfig.scala [hunk 6]
```diff
@@ -4,13 +4,13 @@
...
-import com.amazonaws.ClientConfiguration
-import com.amazonaws.Protocol
 
-class DynamoDBJournalConfig(c: Config) {
+import akka.persistence.dynamodb.{DynamoDBClientConfig, DynamoDBConfig}
+
+
+class DynamoDBJournalConfig(c: Config) extends DynamoDBConfig {
   val JournalTable = c getString "journal-table"
+  val Table = JournalTable
```
Comment (ktoso) [inline src/main/scala/akka/persistence/dynamodb/journal/DynamoDBJournalConfig.scala:13]: why is this needed if it's the same as JournalTable?

File: src/main/scala/akka/persistence/dynamodb/journal/DynamoDBRequests.scala [hunk 7]
```diff
@@ -3,22 +3,90 @@
...
+import scala.concurrent.{ ExecutionContext, Future }
 import scala.concurrent.duration._
 import scala.util.{ Failure, Success, Try }
 import scala.util.control.NonFatal
 import akka.Done
+import akka.actor.{ Actor, ActorLogging }
 import akka.pattern.after
+import akka.persistence.dynamodb._
 
 trait DynamoDBRequests {
```
Comment (ktoso) [inline src/main/scala/akka/persistence/dynamodb/journal/DynamoDBRequests.scala]: Should we add a `@DoNotInherit` here? as the trait is not meant for end users (though I'd hope no one actually tries to extend it)

File: src/test/scala/akka/persistence/dynamodb/snapshot/SnapshotStoreTckSpec.scala [hunk 8]
```diff
@@ -0,0 +1,25 @@
+package akka.persistence.dynamodb.snapshot
```
Comment (ktoso) [inline src/test/scala/akka/persistence/dynamodb/snapshot/SnapshotStoreTckSpec.scala:4]: Please add:

```
/*
 * Copyright (C) 2017 Lightbend Inc. <http://www.lightbend.com>
 */
```

It's Apache v2 but needs this header regardless

File: src/test/scala/akka/persistence/dynamodb/snapshot/SnapshotStoreTckSpec.scala [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+    ensureSnapshotTableExists()
+  }
+
+  override def afterAll(): Unit = {
+    super.afterAll()
+    client.shutdown()
+  }
+}
```
Comment (ktoso) [inline src/test/scala/akka/persistence/dynamodb/snapshot/SnapshotStoreTckSpec.scala:28]: 👍

File: src/test/scala/akka/persistence/dynamodb/snapshot/DynamoDBUtils.scala [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+  private val writerUuid = UUID.randomUUID.toString
+
+  var nextSeqNr = 1
+  def seqNr() = {
+    val ret = nextSeqNr
+    nextSeqNr += 1
+    ret
+  }
+  var generatedMessages: Vector[PersistentRepr] = Vector(null) // we start counting at 1
```
Comment (ktoso) [inline src/test/scala/akka/persistence/dynamodb/snapshot/DynamoDBUtils.scala]: put some dummy value instead of null?

File: src/test/scala/akka/persistence/dynamodb/snapshot/DynamoDBUtils.scala [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+import com.amazonaws.services.dynamodbv2.model._
+
+import scala.collection.JavaConverters._
+import scala.concurrent.{ Await, Future }
+import scala.concurrent.duration._
+import akka.persistence.dynamodb.dynamoClient
+
+trait DynamoDBUtils {
+
+  val system: ActorSystem
```
Comment (ktoso) [inline src/test/scala/akka/persistence/dynamodb/snapshot/DynamoDBUtils.scala]: avoid abstract vals - `def` please

File: src/main/resources/reference.conf [hunk 12]
```diff
@@ -121,3 +121,54 @@ dynamodb-journal {
...
+  # Will only be used if aws-access-key-id is also set.
+  #
+  # It is recommended to leave this (and the aws-access-key-id) setting
+  # empty in order to use the default credentials provider chain, see
+  # http://docs.aws.amazon.com/AWSSdkDocsJava/latest/DeveloperGuide/credentials.html#using-the-default-credential-provider-chain
+  aws-secret-access-key = ${dynamodb-journal.aws-secret-access-key}
+
+  # If this is set to `on` then every DynamoDB request will be logged
+  # at DEBUG level. Caution: this will generate A LOT of output.
+  tracing = ${dynamodb-journal.tracing}
```
Comment (ktoso) [inline src/main/resources/reference.conf:161]: can we rename this to debug perhaps?
as we're working on getting tracing (in the sense of "trails" across multiple components) it could be confusing to users that this perhaps has tracing support already

File: src/main/scala/akka/persistence/dynamodb/DynamoDBConfig.scala [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+  val MaxItemSize: Int
+  val Table: String
+  val JournalName: String
+
+}
+
+trait DynamoDBProvider {
+  val settings: DynamoDBConfig
+  val dynamo: DynamoDBHelper
+  val serialization: Serialization
```
Comment (ktoso) [inline src/main/scala/akka/persistence/dynamodb/DynamoDBConfig.scala]: avoid abstract vals please - make those `def`

**Comment by joost-de-vries** [comment]
@ktoso tx for your code review. I incorporated your comment.

**Comment by joost-de-vries** [comment]
During Scala Days @patriknw mentioned the concern that the snapshot might be too big for dynamodb. Currently there's no special snapshots that are bigger than the Dynamodb maximum item size.
I updated the readme to document the maximum size.

**Comment by joost-de-vries** [comment]
@ktoso I'm wondering what needs to happen to merge this PR? 
I guess my first question is: who do I need to contact? Tx in advance for your help.

**Comment by ktoso** [comment]
Such ping is good enough 😉  Sorry I've forgotten about this PR; I was in the middle of reviewing it actually. Will get back to it and we'll see during the upcoming week - sounds good?

**Comment by joost-de-vries** [comment]
Yeah, that's great.
As I mentioned the things I needed to address implementing snapshot support where mostly:
- how to deal with shared code
- how to deal with shared configurations

I'm actually working on implementing Akka Persistence Query. And that, like the snapshot support, involves shared code. So in a sense it depends on this PR. So I'm glad you'll be reviewing it.

File: README.md [hunk 14]
```diff
@@ -54,6 +55,29 @@ Before you can use these settings you will have to create a table, e.g. using th
...
+~~~
+
+The table to create for snapshot storage has the schema:
+ 
+* a hash key of type String with name `par`
+* a sort key of type Number with name `seq`
+* a sort key of type Number with name `ts`
+* a local secondary index with name `ts-idx` that is an index on the combination of `par` and `ts`
+
+The Dynamodb item of a snapshot [can be 400 kB](http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Limits.html#limits-items). Using a binary serialisation format like ProtoBuf or Kryo will use that space most effectively.
```
Comment (johanandren) [inline README.md:80]: Maybe also hint about compression? (We do that in some serialisers inside of Akka itself)

File: src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+      .map { _ =>
+        ()
+      }
+  }
+
+  def load(persistenceId: String, criteria: SnapshotSelectionCriteria): Future[Option[SelectedSnapshot]] = {
+
+    loadQueryResult(persistenceId, criteria)
+      .map { result =>
+        val results = result.getItems.asScala.map(item => (item.get(Key).getS, item.get(SequenceNr).getN.toLong))
```
Comment (johanandren) [inline src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala]: Never used.

File: src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+        result.getItems.asScala.headOption
+          .map(youngest =>
+            fromSnapshotItem(persistenceId, youngest))
+      }
```
Comment (johanandren) [inline src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala]: Since we are only interested in the head element, maybe save a few allocations here with some imperative code - `if (result.getCount > 0) ...`

File: src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+  }
+
+  private def loadQueryResult(persistenceId: String, criteria: SnapshotSelectionCriteria): Future[QueryResult] = {
+    criteria match {
+      case SnapshotSelectionCriteria(maxSequenceNr, maxTimestamp, minSequenceNr, minTimestamp) if minSequenceNr == 0 && maxSequenceNr == Long.MaxValue =>
+        loadByTimestamp(persistenceId, minTimestamp = minTimestamp, maxTimestamp = maxTimestamp)
+      case SnapshotSelectionCriteria(maxSequenceNr, maxTimestamp, minSequenceNr, minTimestamp) if minTimestamp == 0 && maxTimestamp == Long.MaxValue =>
```
Comment (johanandren) [inline src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala:78]: guards could go directly in pattern, would make it a little bit more clean (not quite sure, let's see what you think):
`case SnapshotSelectionCriteria(maxSequenceNr, maxTimestamp, 0, Long.MaxValue) => ...`

File: src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+
+  private def toSnapshotItem(persistenceId: String, sequenceNr: Long, timestamp: Long, snapshot: Any): Item = {
+    val item: Item = new JHMap
+
+    item.put(Key, S(messagePartitionKey(persistenceId)))
+    item.put(SequenceNr, N(sequenceNr))
+    item.put(Timestamp, N(timestamp))
+    val snp = B(serialization.serialize(Snapshot(snapshot)).get)
+    item.put(Payload, snp)
+    item
```
Comment (johanandren) [inline src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala:137]: Not sure: should we fail fast here if snapshot is too big, given that we know the limit of snapshots?

File: src/main/scala/akka/persistence/dynamodb/DynamoDBConfig.scala [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+}
```
Comment (johanandren) [inline src/main/scala/akka/persistence/dynamodb/DynamoDBConfig.scala]: I'd make all these trait vals into defs and avoid any risk of initialisation order problems. 

Additionally I'm not sure I think serialization belongs here. It's something the `DynamoDBRequests` needs so I'd express it as an abstract method there. 

When I think about it a bit more I think I'd just get rid of this trait entirely and just have settings and dynamo as abstract methods in `DynamoDBRequests` as well, as it would more clearly express that is where the dependency exists. Or am I missing something?

File: src/main/scala/akka/persistence/dynamodb/journal/DynamoDBRequests.scala [hunk 20]
```diff
@@ -3,22 +3,90 @@
...
-  this: DynamoDBJournal =>
+  this: DynamoDBProvider with ActorLogging with Actor =>
```
Comment (johanandren) [inline src/main/scala/akka/persistence/dynamodb/journal/DynamoDBRequests.scala]: As it is now shared between journal and snapshots I think this could be moved one package up.

Maybe just make the whole trait `private[dynamodb]` rather than individual methods

File: src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotStore.scala [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+    save(metadata.persistenceId, metadata.sequenceNr, metadata.timestamp, snapshot)
+  }
+
+  override def deleteAsync(metadata: SnapshotMetadata): Future[Unit] = {
+    delete(metadata)
+  }
+
+  override def deleteAsync(persistenceId: String, criteria: SnapshotSelectionCriteria): Future[Unit] = {
+    delete(persistenceId, criteria)
+  }
```
Comment (johanandren) [inline src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotStore.scala]: Just skip the curlies for these single-expression-methods

File: src/main/scala/akka/persistence/dynamodb/snapshot/package.scala [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+import com.amazonaws.services.dynamodbv2.model._
+
+import scala.collection.generic.CanBuildFrom
+import scala.concurrent._
+import scala.util.{ Failure, Success, Try }
+
+package object snapshot {
+
+  type Item = JMap[String, AttributeValue]
+  type ItemUpdates = JMap[String, AttributeValueUpdate]
```
Comment (johanandren) [inline src/main/scala/akka/persistence/dynamodb/snapshot/package.scala]: These are identical to the ones in the journal, put in the package object one level up?

File: src/main/scala/akka/persistence/dynamodb/snapshot/package.scala [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+      new KeySchemaElement().withAttributeName(Key).withKeyType(KeyType.HASH),
+      new KeySchemaElement().withAttributeName(SequenceNr).withKeyType(KeyType.RANGE)
+    )
+    .withLocalSecondaryIndexes(
+      new LocalSecondaryIndex()
+        .withIndexName(TimestampIndex).withKeySchema(
+          new KeySchemaElement().withAttributeName(Key).withKeyType(KeyType.HASH),
+          new KeySchemaElement().withAttributeName(Timestamp).withKeyType(KeyType.RANGE)
+        ).withProjection(new Projection().withProjectionType(ProjectionType.ALL))
+    )
```
Comment (johanandren) [inline src/main/scala/akka/persistence/dynamodb/snapshot/package.scala]: Just used in the tests as far as I can see so move it there?

File: src/main/scala/akka/persistence/dynamodb/snapshot/package.scala [hunk 24]
```diff
@@ -0,0 +1,81 @@
...
+    p.future
+  }
+
+  def trySequence[A, M[X] <: TraversableOnce[X]](in: M[Future[A]])(implicit
+    cbf: CanBuildFrom[M[Future[A]], Try[A], M[Try[A]]],
+                                                                   executor: ExecutionContext): Future[M[Try[A]]] =
+    in.foldLeft(Future.successful(cbf(in))) { (fr, a) =>
+      val fb = lift(a)
+      for (r <- fr; b <- fb) yield (r += b)
+    }.map(_.result())
```
Comment (johanandren) [inline src/main/scala/akka/persistence/dynamodb/snapshot/package.scala]: These three methods are entirely decoupled from dynamo, but used in both stores, put in a `FutureUtils` or something like that one package up?

File: src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala [hunk 25]
```diff
@@ -0,0 +1,143 @@
...
+  }
+
+  private def fromSnapshotItem(persistenceId: String, item: Item): SelectedSnapshot = {
+    val seqNr = item.get(SequenceNr).getN.toLong
+    val timestamp = item.get(Timestamp).getN.toLong
+    val payloadValue = item.get(Payload).getB
+    serialization.deserialize(payloadValue.array(), classOf[Snapshot])
+      .map(snapshot =>
+        SelectedSnapshot(metadata = SnapshotMetadata(persistenceId, sequenceNr = seqNr, timestamp = timestamp), snapshot = snapshot.data)).get
```
Comment (johanandren) [inline src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala]: just `.get` right away and use the result to construct the `SelectedSnapshot` instead, saves going through a lambda and a try allocation.

File: src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala [hunk 26]
```diff
@@ -0,0 +1,143 @@
...
+import akka.persistence.serialization.Snapshot
+import com.amazonaws.services.dynamodbv2.model._
+import collection.JavaConverters._
+import scala.concurrent.Future
+
+trait DynamoDBSnapshotRequests extends DynamoDBRequests {
+  this: DynamoDBSnapshotStore =>
+
+  import settings._
+  import context.dispatcher
```
Comment (johanandren) [inline src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala:21]: you could cache a `val whateverToUnit = Any => ()` up here and reuse instead of creating one for each place it is used.

In addition to that I think you could use the `akka.dispatch.ExecutionContexts.sameThreadExecutionContext` for those maps, to do that without having to jump into a potentially different thread for essentially a no-op. Maybe make that a helper method in the `FutureUtils` i mentioned somewhere else even - `discardFutureContent(Future[Any]): Future[Unit]`

File: src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala [hunk 27]
```diff
@@ -0,0 +1,143 @@
...
+  }
+
+  def delete(persistenceId: String, criteria: SnapshotSelectionCriteria): Future[Unit] = {
+    for {
+      result <- loadQueryResult(persistenceId, criteria).map(result => result.getItems.asScala.map(item => item.get(SequenceNr).getN.toLong))
+      _ <- doBatch(
+        batch => s"execute batch delete $batch",
+        result.map(snapshotDeleteReq(persistenceId, _))
+      )
+    } yield ()
```
Comment (johanandren) [inline src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala]: This one is a bit messy to read, I think it would be more clear with (nicely line breaks and) map/flatMap all the way.

**Review by johanandren — COMMENTED** [review]
Did a review. I don't have enough Dynamodb knowledge to say if the way it uses dynamo is sound, so feedback is mostly about style and more Akka/concurrency-related things.

Looks very promising, great work!

File: README.md (hunk 14 above)
Comment (joost-de-vries) [reply README.md:80]: Do you mean: use compression in your serialisers if you're using json as a persistence format?

File: src/main/scala/akka/persistence/dynamodb/DynamoDBConfig.scala (hunk 19 above)
Comment (joost-de-vries) [reply src/main/scala/akka/persistence/dynamodb/DynamoDBConfig.scala]: I removed the dynamodbprovider trait. The settings I couldn't turn into a def because of the convention in this codebase of importing it, and that requires a stable value.

File: src/main/scala/akka/persistence/dynamodb/journal/DynamoDBRequests.scala (hunk 20 above)
Comment (joost-de-vries) [reply src/main/scala/akka/persistence/dynamodb/journal/DynamoDBRequests.scala]: Done

File: src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala (hunk 27 above)
Comment (joost-de-vries) [reply src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala]: refactored

File: src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala (hunk 15 above)
Comment (joost-de-vries) [reply src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala]: removed

File: src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala (hunk 16 above)
Comment (joost-de-vries) [reply src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala]: Done

File: src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala (hunk 17 above)
Comment (joost-de-vries) [reply src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala:78]: The order of those parameters are a bit unexpected. So checking their value by name makes it clearer what's going on IMO.

File: src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala (hunk 18 above)
Comment (joost-de-vries) [reply src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala:137]: Unfortunately Dynamodbs way of determining the size is a bit complicated. So I'd rather not try to duplicate that logic.

File: src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala (hunk 25 above)
Comment (joost-de-vries) [reply src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala]: Done

File: src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotStore.scala (hunk 21 above)
Comment (joost-de-vries) [reply src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotStore.scala]: Done

File: src/main/scala/akka/persistence/dynamodb/snapshot/package.scala (hunk 22 above)
Comment (joost-de-vries) [reply src/main/scala/akka/persistence/dynamodb/snapshot/package.scala]: Done

File: src/main/scala/akka/persistence/dynamodb/snapshot/package.scala (hunk 23 above)
Comment (joost-de-vries) [reply src/main/scala/akka/persistence/dynamodb/snapshot/package.scala]: Done

File: src/main/scala/akka/persistence/dynamodb/snapshot/package.scala (hunk 24 above)
Comment (joost-de-vries) [reply src/main/scala/akka/persistence/dynamodb/snapshot/package.scala]: moved to dynamodb package

File: src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala (hunk 26 above)
Comment (joost-de-vries) [reply src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala:21]: I've created a val as you said.

Wrt sameThreadExecutionContext: TBH I haven't used that before. What would be the proper way of using that?

**Comment by joost-de-vries** [comment]
Hi Johan, tx for your review.

I've fixed your comments. With 2 questions.

Wrt the Dynamodb stuff:
Dynamodb is partitioned. And if you overload a single partition you get throttled. The journal implementation shards writes of a single persistenceid further to prevent write overloading.
For the journal I did not do a similar mechanism since I think typically a single snapshot at a time is made for a persistenceid.

Another design choice was how to query snapshots. The selectioncriteria can be by sequence or timestamp. I've added a local secondary index on partitionKey + timestamp. 
If the criteria search for either seq or timestamp I can read them optimally without consuming unnecessarily capacity.
I don't see how to achieve that if both sequence nr and timestamp are selected; dynamodb supports only using one index at a time.

**Review by ktoso — COMMENTED** [review]
Looking good, let's polish up and try to merge this week - thanks for your continued effort here :)

**Comment by johanandren** [comment]
Sharding snapshots on id sounds good.

With regards to querying, the most important part is that querying for `SnapshotSeletionCriteria.Latest` is efficient since that is what persistence normally does on every actor wakeup. I think that all the other ways of custom querying for snapshots could be slow/expensive/possibly even not implemented and just clearly documented that it will be bad to do.

File: src/main/resources/reference.conf (hunk 3 above)
Comment (joost-de-vries) [reply src/main/resources/reference.conf]: Done

File: src/main/resources/reference.conf (hunk 12 above)
Comment (joost-de-vries) [reply src/main/resources/reference.conf:161]: This was existing functionality & name from the journal implementation. We'll have to rename it in the journal implementation as well. With possible problems for existing users... Shall we leave it as is for now?

File: src/main/resources/reference.conf (hunk 2 above)
Comment (joost-de-vries) [reply src/main/resources/reference.conf:171]: Done

File: src/main/scala/akka/persistence/dynamodb/journal/DynamoDBJournalConfig.scala (hunk 6 above)
Comment (joost-de-vries) [reply src/main/scala/akka/persistence/dynamodb/journal/DynamoDBJournalConfig.scala:13]: fixed

File: src/test/scala/akka/persistence/dynamodb/snapshot/SnapshotStoreTckSpec.scala (hunk 8 above)
Comment (joost-de-vries) [reply src/test/scala/akka/persistence/dynamodb/snapshot/SnapshotStoreTckSpec.scala:4]: done

File: src/test/scala/akka/persistence/dynamodb/snapshot/DynamoDBUtils.scala (hunk 11 above)
Comment (joost-de-vries) [reply src/test/scala/akka/persistence/dynamodb/snapshot/DynamoDBUtils.scala]: done

File: src/test/scala/akka/persistence/dynamodb/snapshot/DynamoDBUtils.scala (hunk 10 above)
Comment (joost-de-vries) [reply src/test/scala/akka/persistence/dynamodb/snapshot/DynamoDBUtils.scala]: removed

File: README.md (hunk 5 above)
Comment (joost-de-vries) [reply README.md]: Nice catch. I clarified it.

**Comment by joost-de-vries** [comment]
@johanandren That's a good point wrt the most used query. `SnapshotSeletionCriteria.Latest` uses the min max for both seq and ts and that will be executed using the standard query by partitionkey and seq. Which is more capacity (cost) efficient than if it used the timestamp index. So that's good.

I changed that the queries are using [consistent reads](http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.ReadConsistency.html). That consumes twice the capacity and fails if the network between availability zones fails. 
The journal queries use consistent reads everywhere as well.

@ktoso Konrad, tx for your review. I incorporated all your comments.

File: src/main/resources/reference.conf (hunk 12 above)
Comment (ktoso) [reply src/main/resources/reference.conf:161]: Yeah makes sense to keep then 👍

File: src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala (hunk 26 above)
Comment (ktoso) [reply src/main/scala/akka/persistence/dynamodb/snapshot/DynamoDBSnapshotRequests.scala:21]: Not a biggie; can get in with the normal one.
That context can be imported and used like any other one, but it hijacks the calling thread to execute the things.

File: src/test/scala/akka/persistence/dynamodb/snapshot/SnapshotStoreTckSpec.scala (hunk 8 above)
Comment (ktoso) [reply src/test/scala/akka/persistence/dynamodb/snapshot/SnapshotStoreTckSpec.scala:4]: Thanks a ton :)

**Comment by ktoso** [comment]
Thanks a ton and sorry for the large delays in reviewing this - vacation season hit the team hard but we're back in pretty good capacity finally :)

I'm going to merge this and release an 1.1 so this can be used right away.

Thanks again for your persistence (pun intended) and work on this!
